### ISSUE TITLE
Add floppy write support

### ISSUE THREAD

**Issue by mikrosk** [atarijookie/ce-atari#156]
If I run CE_FDD.PRG, press "N" to create a new disk, I can open it in the desktop. However, if I write a file to A:, nothing is shown (copy operation ends with success but that's it).

**Comment by atarijookie** [comment]
Write support not done (yet) on floppy.

**Comment by mikrosk** [comment]
Excuse my ignorance but then what is the new disk good for? :-)

**Comment by atarijookie** [comment]
> Excuse my ignorance but then what is the new disk good for? :-)

For the future write support :) You want to help with that one?

**Comment by atarijookie** [comment]
The write support needs first the #183 to be done - if the floppy image will be downloaded and inserted on native (RAW) drive, writing to it from ce_main_app would be problematic, so for most of the cases downloading, inserting and writing from ImageStorage (USB drive or shared drive) will be the way to go.

**Comment by atarijookie** [comment]
TODO:
- [x] 1) alter encoding in imageSilo to work on tracks and allow access to tracks sooner than all the tracks are encoded (kick out copyFromOther)
- [x] 2) if the track in imageSilo isn't encoded yet, force floppy encoder to encode that track first and then let it continue with the remaining tracks
- [x] 3) alter encoding order algorithm - instead of adding image to queue, let encoder look at currently selected slot and if it's not fully encoded, let it encode that one (user will probably use that one now), and if there's nothing to be done in currently selected slot let it look in other slots and encode them then and then let it sleep 
- [x] 4) wake up encoder on add new image to slot or reencode track request
- [x] 5) make sure that the floppy image loaded from translated drive isn't copied to /tmp (we would loose data then)
```slots[positionIndex].hostDestPath   = hostDestPath;     // where the file is stored when used: /tmp/bla.st```
- [x] 6) connect CCoreThread::handleSectorWritten to floppy encoder
- [x] 7) add / check / fix decoding of new MFM stream
- [x] 8) add writing of the new sector to FloppyImage
- [x] 9) force the encoder to re-encode the track which just received new data
- [x] 10) send new track to Franz
- [x] 11) save the modified floppy image to translated drive - if last sector was written at least 5 seconds ago, do image write (to not write too much after every sector write)
- [x] 12) image write on encode thread quit
- [x] 13) image write on load new image and something not saved in old image
- [x] 14) refactoring of floppyImage - to open-load-close, isOpen -> isLoaded, open-save-close
- [x] 15) add code for saving .ST images (now it's just empty)
- [x] 16) test / fix write operation - might need changes in Franz
- [x] 17) apply read-only / read-write setting of floppy (avoid overwriting in read-only mode)
- [x] 18) make the default empty image always read-only 

...okay, now this seems like much more work than I expected :-/

**Comment by atarijookie** [comment]
I've made changes for tasks 1), 2), 3), 4), so later I can easily do 9), but I've completely made a mess there, not yet tested, needs a lots of improvements until I can say that 1-4 are done 👎

**Comment by atarijookie** [comment]
Things TODO next:
- [x] split encoder thread and addEncode...() functions to separate file - refactoring (imagesilo functions vs going through slots and encoding tracks)
- [x] move loading of image file from addEncodeWholeImageRequest() to encoding thread, so addEncodeWholeImageRequest() will take as least as it could
- [x] protect corethread vs encodethread access to imagesilo by mutex, but don't make the encode thread or core wait too much (encoding can take seconds) -- that should be easier to to with the floppy encoder in a separate file

**Comment by atarijookie** [comment]
- [x] Each encoded track should have two DWORDs associated:
- Encode start - written by encoder thread when it starts ti encode
- encode request - written by core thread after write - on reencode request

When encoder finishes encoding that track, he should look and compare times, and if start time < request time, there was another reencode request while it was encoding that track, so it should reencode it again...

- [x] Same could be used for open image request vs open image time by encode thread.

**Comment by atarijookie** [comment]
Most of the code is there, but nothing was tested since I've done all the refactoring on reading and added writing, so the plan now is:
- [x] test / fix floppy reading
- [x] test / fix / improve floppy writing

**Comment by atarijookie** [comment]
The floppy read now hangs a bit due to encoding / badly marked flags:
```
00000003 (2018-12-01 05:10:42.397006)  ATN_SEND_TRACK -- Franz wants: [track 0, side 0]. Current image has: [track 80, side 2, sectors/track: 9]
00000000 (2018-12-01 05:10:42.397166)  floppyEncoder_addReencodeTrackRequest - track: 0, side: 0
00000010 (2018-12-01 05:10:42.407186)  Encoding of [track 0, side 0] of image /tmp/emptyimage.st done
00000010 (2018-12-01 05:10:42.417094)  Encoding of [track 0, side 0] of image /tmp/emptyimage.st done
00000488 (2018-12-01 05:10:42.905502)  >>> Suspicious core cycle load -- cycle time: 1179 ms, load:  43 %
```

**Comment by atarijookie** [comment]
Floppy reading tested and fixed and reading works fine again. With improved encoding speed :) Possibly we could also skip waiting in web and ST apps for encoding to finish completely as the encoding now works per individual tracks and prioritizes track if it's needed for reading but it's not encoded yet.

**Comment by atarijookie** [comment]
```
04682096 00000440 (2018-12-04 04:29:45.527540)	handleSectorWritten -- track 2, side 0, sector 1
04682096: outBfr - 806 bytes
        01 02 00 55 55 55 15 55 15 55 15 55 15 55 15 55 | ...UUU.U.U.U.U.U
        15 55 15 55 15 55 15 55 55 55 05 5b 39 e9 3a 55 | .U.U.U.UUU.[9.:U
        16 55 15 55 19 56 25 59 25 a5 16 e5 26 55 05 96 | .U.U.V%Y%...&U..
        15 5a 25 b5 1b 56 19 59 39 96 15 56 19 55 26 9a | .Z%..V.Y9..V.U&.
        09 6a 26 e5 2f 6d 26 d6 16 55 06 a6 39 65 26 55 | .j&./m&..U..9e&U
        19 66 15 59 15 a5 15 96 29 a6 06 6a 25 ae 2a 9a | .f.Y....)..j%.*.
        26 b9 06 ea 3e 7d 29 96 07 a5 26 d6 26 56 19 a6 | &...>})...&.&V..
        26 69 09 e5 1f ad 26 56 25 56 16 a5 09 97 29 59 | &i....&V%V....)Y
        25 79 15 95 15 d5 16 96 39 aa 15 e6 06 9b 26 69 | %y......9.....&i
        25 a9 1a ae 0e ad 3a 9a 27 a5 0a 95 39 5e 1b aa | %.....:.'...9^..
        0a 7e 06 ff 1a f7 36 6d 0a de 37 75 26 5f 05 6b | .~....6m..7u&_.k
        15 d9 16 65 36 69 26 5a 25 af 0a 69 25 d9 19 ad | ...e6i&Z%..i%...
        0e 9d 37 69 37 75 27 57 19 59 09 b5 2a 59 35 d9 | ..7i7u'W.Y..*Y5.
        1f 5d 09 75 15 95 16 b5 3a 5d 35 66 15 f5 16 95 | .].u....:]5f....
        15 95 16 55 25 5b 15 96 06 a9 2d 95 39 55 26 69 | ...U%[....-.9U&i
        29 9b 09 b5 19 a6 1d a5 39 55 29 6a 0a 6a 3a a6 | ).......9U)j.j:.
        2e af 06 de 35 66 1a 69 2b e9 05 a5 2a 56 25 69 | ....5f.i+...*V%i
        15 65 1a 59 0d bd 25 ea 3a 7a 0f a5 3e 7e 09 ff | .e.Y..%.:z..>~..
        35 da 0d df 1f d5 36 6d 25 da 0d a5 1e 9f 1f 7d | 5.....6m%......}
        05 75 1b 59 35 f5 37 56 09 5d 15 75 15 59 16 65 | .u.Y5.7V.].u.Y.e
        25 97 19 55 26 99 2d 99 39 95 06 99 2a 5b 36 d5 | %..U&.-.9...*[6.
        2a 9a 15 a5 06 95 1e 5b 26 5a 36 95 0e 9f 17 77 | *......[&Z6....w
        35 76 05 da 1d fd 35 56 25 75 35 dd 15 59 19 65 | 5v....5V%u5..Y.e
        05 66 16 65 1a a6 2d 69 35 95 16 65 0e 57 35 f5 | .f.e..-i5..e.W5.
        1d 97 1d 75 37 55 16 59 19 95 1b 56 05 57 25 75 | ...u7U.Y...V.W%u
        17 55 35 55 26 56 25 5f 15 d5 15 a5 17 55 17 55 | .U5U&V%_.....U.U
        16 55 15 55 19 56 25 59 25 a5 16 e5 16 55 16 5a | .U.U.V%Y%....U.Z
        05 5a 25 b9 1b 66 19 59 39 96 15 56 19 55 2a 9a | .Z%..f.Y9..V.U*.
        09 6a 26 e5 2f ad 25 d6 16 55 2a 99 09 65 26 55 | .j&./.%..U*..e&U
        19 76 15 99 15 a5 15 96 29 a6 0a 6a 25 aa 2a 5a | .v......)..j%.*Z
        25 b9 06 ea 3e 7d 29 96 07 ad 26 d6 26 56 19 a6 | %...>})...&.&V..
        2a 69 27 96 0f a5 26 56 25 56 16 a5 36 5d 25 65 | *i'...&V%V..6]%e
        05 79 15 99 15 e5 16 96 39 6a 15 e6 0a 9b 26 69 | .y......9j....&i
        25 e9 1a ae 0e a9 3a 5a 27 a5 0a 95 39 5e 1b aa | %.....:Z'...9^..
        0a 7e 1b ff 0a f7 36 6d 2b 7a 07 75 26 5f 35 ad | .~....6m+z.u&_5.
        17 65 06 65 36 69 26 5a 26 af 06 69 25 d9 19 ed | .e.e6i&Z&..i%...
        0e 9f 37 65 37 75 2b 57 19 59 05 b5 2a 59 35 d9 | ..7e7u+W.Y..*Y5.
        1f 5d 09 75 15 96 16 b5 3a 5d 35 66 15 75 15 95 | .].u....:]5f.u..
        15 95 16 55 25 5b 25 96 1a a5 0d 95 39 65 26 a9 | ...U%[%.....9e&.
        2a 9b 0d b5 19 aa 1d 95 39 55 2a 6a 25 aa 0a a6 | *.......9U*j%...
        2e af 06 de 35 66 1a a9 29 e9 16 96 0a 5e 25 79 | ....5f..)....^%y
        15 65 1b 59 0d be 25 ea 3a 7a 07 a5 3e 6e 09 ff | .e.Y..%.:z..>n..
        35 da 3f 7d 0f d5 36 5d 26 da 0d a5 1e bf 1d 7d | 5.?}..6]&......}
        05 75 1b 59 35 f5 35 56 09 5d 15 7d 15 59 16 65 | .u.Y5.5V.].}.Y.e
        25 97 19 65 26 99 2d 99 35 95 1a 66 0a 5b 36 95 | %..e&.-.5..f.[6.
        29 9a 15 a5 1a 55 0e 59 26 5a 37 95 0e 9f 17 f7 | )....U.Y&Z7.....
        35 76 17 69 0d f5 35 56 27 75 35 dd 15 59 19 65 | 5v.i..5V'u5..Y.e
        05 66 16 65 1a a6 2d 59 35 95 17 65 39 5f 05 f5 | .f.e..-Y5..e9_..
        1d 97 1d 75 37 55 16 59 15 95 1b 56 05 57 25 7d | ...u7U.Y...V.W%}
        17 75 35 55 26 56 25 5f 15 d5 15 a5 15 55 15 55 | .u5U&V%_.....U.U
        15 a7 0d e9 00 00                               | ......          
04682098 00000002 (2018-12-04 04:29:45.529662)	floppyEncoder_decodeMfmWrittenSector - track: 2, side: 0, sector: 1, size: 806 -> writtenSector index: 0
04682109 00000011 (2018-12-04 04:29:45.540249)	MfmCachedImage::decodeMfmBuffer - sync not found
04682109 00000000 (2018-12-04 04:29:45.540436)	Written sector at index: 0 decoded, good: 0


04677298 00000641 (2018-12-04 04:29:40.729347)	handleSectorWritten -- track 2, side 0, sector 1
04677298: outBfr - 806 bytes
        01 02 05 55 15 55 15 55 15 55 15 55 15 55 15 55 | ...U.U.U.U.U.U.U
        15 55 55 55 15 55 15 55 15 55 15 5b 0a 7b 1e d5 | .UUU.U.U.U.[.{..
        17 55 15 55 16 95 29 56 19 69 15 79 16 95 19 96 | .U.U..)V.i.y....
        25 5a 25 b9 06 d9 26 55 1e 55 25 65 26 95 1a 66 | %Z%...&U.U%e&..f
        29 6a 16 e5 0b e7 19 75 26 95 19 a6 19 67 09 96 | )j.....u&....g..
        16 5d 25 66 15 69 15 65 2a a9 29 6a 09 6a 2a 96 | .]%f.i.e*.)j.j*.
        29 6e 15 ea 0b 9b 1a 65 2d ad 09 b5 29 95 26 69 | )n.....e-...).&i
        2a 5a 1a e5 09 e9 19 9d 29 65 25 a9 1e 97 19 59 | *Z......)e%....Y
        15 7a 15 99 05 79 15 a5 2e 5a 25 99 29 9b 09 9a | .z...y...Z%.)...
        19 7a 16 ab 0a aa 1e 96 29 e9 1a 95 0e 56 26 ea | .z......)....V&.
        0b 9f 26 ff 36 f5 0d 97 1a de 09 df 19 b7 3d 6b | ..&.6.........=k
        15 d9 16 65 09 9a 19 9e 29 9b 36 69 09 77 16 7b | ...e....).6i.w.{
        1f 9f 05 d9 1d fd 1a 55 35 56 15 b5 2a 5a 0d 7a | .......U5V..*Z.z
        17 d7 19 75 35 96 16 b5 0e 95 1d 55 25 5d 15 65 | ...u5......U%].e
        15 55 15 95 19 96 39 65 25 a9 1d 96 0e 59 19 a6 | .U....9e%....Y..
        1a 96 3d b5 29 a9 07 65 1e 65 1a 9a 29 6a 2a a5 | ..=.)..e.e..)j*.
        0b ab 36 de 25 69 06 a6 1a 7a 1d a5 2a 5d 09 5e | ..6.%i...z..*].^
        15 5d 16 d6 1e be 05 7a 2e 9e 27 a5 07 9b 29 ff | .].....z..'...).
        0d 7e 2f 5f 17 d5 0d 96 19 96 2d a5 0f ad 37 5f | .~/_......-...7_
        1d 75 2b 5a 0d 7d 1d 55 2b 5d 35 7d 15 59 16 65 | .u+Z.}.U+]5}.Y.e
        15 96 06 59 19 aa 1b 56 1d 65 16 99 2a 5a 0d a5 | ...Y...V.e..*Z..
        1a 66 27 69 1e 95 16 59 09 97 2d e5 1f 9f 05 f5 | .f'i...Y..-.....
        3d 5d 26 da 05 7d 1d 5d 29 dd 1d 77 15 56 15 59 | =]&..}.])..w.V.Y
        19 66 36 65 1a a5 0b 56 1d 65 15 d9 1e 57 35 f5 | .f6e...V.e...W5.
        07 69 37 55 1d 55 15 56 15 a5 16 d5 29 57 25 7d | .i7U.U.V....)W%}
        05 dd 1d 55 19 95 29 55 35 55 15 59 15 55 15 55 | ...U..)U5U.Y.U.U
        15 65 15 55 16 95 29 56 19 59 15 69 16 55 19 96 | .e.U..)V.Y.i.U..
        25 5a 29 b9 1b 65 06 55 1e 59 25 65 26 95 1a 66 | %Z)..e.U.Y%e&..f
        2a 6a 1a e5 07 e7 19 65 26 55 19 a6 1d 67 16 59 | *j.....e&U...g.Y
        06 5e 25 65 15 69 15 69 2a 99 29 6a 25 ab 0a 9a | .^%e.i.i*.)j%...
        29 6e 15 aa 0b 9b 1a e5 2d ad 16 d5 26 55 06 69 | )n......-...&U.i
        2a 6a 1a 65 09 e9 19 9d 29 65 25 a9 1e 57 19 59 | *j.e....)e%..W.Y
        19 7a 15 99 15 e5 16 95 0e 5a 25 99 29 9b 36 6a | .z.......Z%.).6j
        09 7a 16 ab 2a ab 0e 9a 29 e9 1a 95 0e 56 26 fa | .z..*...)....V&.
        2e 7f 06 ff 36 f5 0d 96 1a de 09 df 19 b7 3d 6b | ....6.........=k
        15 d9 15 65 26 6a 09 9d 29 9b 36 69 25 de 06 7b | ...e&j..).6i%..{
        1f 9f 17 66 0d f5 1a 55 35 56 16 b5 2a 5a 05 79 | ...f...U5V..*Z.y
        17 d7 1a 75 35 96 16 b5 06 95 1d 55 25 5d 15 65 | ...u5......U%].e
        15 55 15 a5 19 96 39 65 25 a9 19 96 19 65 09 a6 | .U....9e%....e..
        1a 96 3e b5 25 a9 07 66 1e 65 1a 96 29 aa 2e a5 | ..>.%..f.e..)...
        0b ab 36 9e 25 69 06 a6 1a 7a 1d a5 2a 5d 09 5e | ..6.%i...z..*].^
        15 5d 16 e6 1e 7e 15 ea 0e be 27 a5 07 9b 29 ff | .]...~....'...).
        0d 7d 2f 5f 17 d5 05 96 19 96 2f a5 0d ad 37 7f | .}/_....../...7.
        1d 75 27 5a 05 7d 1d 55 2b 5d 35 7d 15 59 16 65 | .u'Z.}.U+]5}.Y.e
        15 96 19 65 09 a9 1b 56 1d a5 16 99 2a 5a 0d ad | ...e...V....*Z..
        1a 66 27 69 1d 95 16 59 0d 97 2d e5 1f 9f 05 f5 | .f'i...Y..-.....
        3d 5d 26 5a 15 f7 0d 5d 29 d5 1d 57 15 56 15 59 | =]&Z...])..W.V.Y
        19 66 36 65 16 a5 0b 56 1d 69 15 e9 1e 97 3d f5 | .f6e...V.i....=.
        07 69 37 55 1d 55 15 66 15 95 16 95 29 57 29 7d | .i7U.U.f....)W)}
        17 75 35 55 09 95 29 55 35 55 15 59 15 75 15 75 | .u5U..)U5U.Y.u.u
        15 59 3f e9 00 00                               | .Y?...          
04677300 00000002 (2018-12-04 04:29:40.731614)	floppyEncoder_decodeMfmWrittenSector - track: 2, side: 0, sector: 1, size: 806 -> writtenSector index: 0
04677311 00000011 (2018-12-04 04:29:40.742149)	MfmCachedImage::decodeMfmBuffer - sync not found
04677311 00000000 (2018-12-04 04:29:40.742314)	Written sector at index: 0 decoded, good: 0
```

- [x] track / sector numbers are swapped on write - is track 2, sector 1, should be sector 2, track 1
- [x] code fails to find A1 sync

**Comment by atarijookie** [comment]
Okay, so the decoder can't find the sync word in the written stream... Reasons and possible improvements:
- [x] check floppy write signal with scope
- [x] if the scope shows longer pulses instead of short ones like when doing read, then there will be needed changes of capturing stream - it seems that timer can capture either falling or rising edge, but not both
- [x] check / do byte swapping (was added in decoder, not sure if needed)
- [x] overflow of mfmWriteStreamBuffer[] when doing getMfmWriteTimes() in Franz - check, increase buffer size from 8 to 16/32/64?
- [x] when doing ```if(tmp < 20) { // too short? skip it``` then maybe should add that value to the next time, because that might change detected pulse from 4 us to 6 us, from 6 us to 8 us or so... 
- [x] also check for happening of ```} else { // too long?``` - might show some other issue, too
- [ ] possibly send unprocessed captured times to RPi for analysis

**Comment by atarijookie** [comment]
So the floppy data signal is made of the shorter pulses of same polarity, which can be captured on the falling edge of the signal...
![scope_0](https://user-images.githubusercontent.com/18427437/49560107-76effe00-f911-11e8-8c41-e6d5821b6fa2.png)

**Comment by atarijookie** [comment]
Even though I can't get debugging on Franz going :( the chip managed to send some sector data to RPi, which now looks better:
```04976531 00000624 (2018-12-06 04:52:34.459746)  handleSectorWritten -- track 0, side 0, sector 1
04976532: outBfr - 808 bytes
        00 01 00 b6 55 55 55 55 55 55 55 55 55 55 55 55 | ....UUUUUUUUUUUU
        55 55 55 55 55 55 55 55 55 55 55 5b b9 ee 7b 95 | UUUUUUUUUUU[..{.
        5d 95 55 55 5a 55 a5 56 65 69 55 b9 59 95 65 96 | ].UUZU.VeiU.Y.e.
        95 5a a5 b9 5b 65 99 56 79 65 95 95 9a 55 69 a6 | .Z..[e.Vye...Ui.
        a9 6a 66 e5 6f 9b 65 b5 99 95 66 a6 79 67 66 59 | .jf.o.e...f.ygfY
        59 79 95 96 55 a9 55 a5 aa 69 a6 6a a5 ab aa 66 | Yy..U.U..i.j...f
        a5 ae 56 ea be 6f 6b 65 b7 ad 66 d5 a6 55 99 a9 | ..V..oke..f..U..
        a9 9a 69 e5 9f a7 66 75 a5 95 96 a9 79 97 69 59 | ..i...fu....y.iY
        65 7a 55 99 55 e5 56 95 b9 6a 96 79 a6 9b e6 6a | ezU.U.V..j.y...j
        65 ea 5a ab ae ab 7a 66 a7 e9 6a 95 f9 5b 9b ea | e.Z...zf..j..[..
        ba 7f 9b ff da f5 f6 5b 6b de b7 7d 66 d7 f5 6b | .......[k..}f..k
        55 d9 56 65 b6 6a 66 76 a6 6b da 69 a5 de 59 eb | U.Ve.jfv.k.i..Y.
        7e 9f 77 66 77 dd 69 d5 d6 56 59 b5 aa 5a 75 e6 | ~.wfw.i..VY..Zu.
        5f d7 69 75 d5 96 56 b5 7a 57 75 59 95 7d 55 a5 | _.iu..V.zWuY.}U.
        55 65 56 95 66 56 e5 65 96 a9 6d 96 79 65 66 9a | UeV.fV.e..m.yef.
        6a 66 f9 b5 99 a9 9d 99 79 95 6a 5a a6 6a ba a5 | jf......y.jZ.j..
        ae ab da de b5 69 9a 9a 69 fa 75 a5 aa 5d a5 7a | .....i..i.u..].z
        55 79 5b 96 79 be 65 ea ba de 9f a5 7e 6f a7 ff | Uy[.y.e.....~o..
        f5 f6 bd df 5f d5 76 5b 66 76 bd a5 de b7 dd df | ...._.v[fv......
        75 75 9b 5a 75 fd 75 d5 ad 5d d5 7d 55 59 56 65 | uu.Zu.u..].}UYVe
        65 96 d9 65 66 a6 6d 66 76 65 5a 99 aa 5a f6 b5 | e..ef.mfveZ..Z..
        69 a6 9d 69 76 95 5e 59 e6 5e b7 a5 7e 9f d7 dd | i..iv.^Y.^..~...
        f5 5d 99 da 5d f7 75 75 a7 5d 75 77 55 56 56 59 | .]..].uu.]uwUVVY
        65 66 d6 65 5a a5 ad 5a 75 a5 57 99 7a 57 f5 f5 | ef.eZ..Zu.W.zW..
        5d a5 dd 5d 75 d5 55 96 56 65 5a d5 a5 57 a5 7d | ]..]u.U.VeZ..W.}
        57 75 75 55 66 55 a5 57 d5 75 55 69 55 d5 55 d5 | WuuUfU.W.uUiU.U.
        55 95 55 55 5a 55 a5 56 65 69 55 b9 59 95 65 96 | U.UUZU.VeiU.Y.e.
        95 5a a5 b9 5b 65 99 56 79 65 95 95 9a 55 69 a6 | .Z..[e.Vye...Ui.
        a9 6a 66 e5 6f 9b 65 b5 99 95 66 a6 79 67 66 59 | .jf.o.e...f.ygfY
        59 79 95 96 55 a9 55 a5 aa 69 a6 6a a5 ab aa 66 | Yy..U.U..i.j...f
        a5 ae 56 ea be 6f 6b 65 b7 ad 66 d5 a6 55 99 a9 | ..V..oke..f..U..
        a9 9a 69 e5 9f a7 66 75 a5 95 96 a9 79 97 69 59 | ..i...fu....y.iY
        65 7a 55 99 55 e5 56 95 b9 6a 96 79 a6 9b e6 6a | ezU.U.V..j.y...j
        65 ea 5a ab ae ab 7a 66 a7 e9 6a 95 f9 5b 9b ea | e.Z...zf..j..[..
        ba 7f 9b ff da f5 f6 5b 6b de b7 7d 66 d7 f5 6b | .......[k..}f..k
        55 d9 56 65 b6 6a 66 76 a6 6b da 69 a5 de 59 eb | U.Ve.jfv.k.i..Y.
        7e 9f 77 66 77 dd 69 d5 d6 56 59 b5 aa 5a 75 e6 | ~.wfw.i..VY..Zu.
        5f d7 69 75 d5 96 56 b5 7a 57 75 59 95 7d 55 a5 | _.iu..V.zWuY.}U.
        55 65 56 95 66 56 e5 65 96 a9 6d 96 79 65 66 9a | UeV.fV.e..m.yef.
        6a 66 f9 b5 99 a9 9d 99 79 95 6a 5a a6 6a ba a5 | jf......y.jZ.j..
        ae ab da de b5 69 9a 9a 69 fa 75 a5 aa 5d a5 7a | .....i..i.u..].z
        55 79 5b 96 79 be 65 ea ba de 9f a5 7e 6f a7 ff | Uy[.y.e.....~o..
        f5 f6 bd df 5f d5 76 5b 66 76 bd a5 de b7 dd df | ...._.v[fv......
        75 75 9b 5a 75 fd 75 d5 ad 5d d5 7d 55 59 56 65 | uu.Zu.u..].}UYVe
        65 96 d9 65 66 a6 6d 66 76 65 5a 99 aa 5a f6 b5 | e..ef.mfveZ..Z..
        69 a6 9d 69 76 95 5e 59 e6 5e b7 a5 7e 9f d7 dd | i..iv.^Y.^..~...
        f5 5d 99 da 5d f7 75 75 a7 5d 75 77 55 56 56 59 | .]..].uu.]uwUVVY
        65 66 d6 65 5a a5 ad 5a 75 a5 57 99 7a 57 f5 f5 | ef.eZ..Zu.W.zW..
        5d a5 dd 5d 75 d5 55 96 56 65 5a d5 a5 57 a5 7d | ]..]u.U.VeZ..W.}
        57 75 75 55 66 55 a5 57 d5 75 55 69 55 d5 55 d5 | WuuUfU.W.uUiU.U.
        55 69 fd e9 a5 55 00 00                         | Ui...U..
04976534 00000003 (2018-12-06 04:52:34.462640)  floppyEncoder_decodeMfmWrittenSector - track: 0, side: 0, sector: 1, size: 808 -> writtenSector index: 0
04976545 00000011 (2018-12-06 04:52:34.473490)  MfmCachedImage::decodeMfmBuffer - sync not found
04976545 00000000 (2018-12-06 04:52:34.473735)  Written sector at index: 0 decoded, good: 0
```

- [x] plus the test tool seems to write only once after starting, have to check that out, verify with scope
- [x] kick out geometry detection on write single sector test

**Comment by atarijookie** [comment]
I've tested the decoding on PC, the data up there seem to be valid! I was able to decode it, even received and calculated CRC matched. Byte swapping on host not needed, so I should remove it (```// swap bytes - Franz has other endiannes```).

**Comment by atarijookie** [comment]
Code changed a bit, written and decoded sector seems to be fine, so issues which remain are:
- [x] don't slot->encImage.askToReencodeTrack immediatelly, but instead after the written sector is decoded
- [x] image file saving after sector write doesn't seem to work
- [x] the reencoded track is not sent back to Franz for immediate streaming, so the write-read test fails -- add code to Franz that when a sector was written during one media rotation, ask for new track data
- [x] extend test to write more than 1 sector at the time!

```
01603696 00000467 (2018-12-11 05:07:59.408948)  handleSectorWritten -- track 2, side 1, sector 3
01603696 00000000 (2018-12-11 05:07:59.409280)  floppyEncoder_decodeMfmWrittenSector - track: 2, side: 1, sector: 3, size: 808 -> writtenSector index: 0
01603697 00000001 (2018-12-11 05:07:59.410609)  MfmCachedImage::handleDecodedByte - received CRC: 3b6, calculated CRC: 3b6, good: 1
01603698 00000001 (2018-12-11 05:07:59.410902)  Written sector at index: 0 decoded, good: 1
01603698 00000000 (2018-12-11 05:07:59.411206)  FloppyImage::readNotWriteSector -- sectorsWritten: 1, lastWriteTime: 00187872
01603708 00000010 (2018-12-11 05:07:59.421289)  Encoding of [slot: 0, track 2, side 1] of image /tmp/newimg0.st done
01603835 00000127 (2018-12-11 05:07:59.548756)  FW: Franz, 2018-12-05
01604229 00000394 (2018-12-11 05:07:59.942302)  FW: Hans,  2017-04-13, LED is: 0, XI: 0x21
01604836 00000607 (2018-12-11 05:08:00.549312)  FW: Franz, 2018-12-05
01605230 00000394 (2018-12-11 05:08:00.943033)  FW: Hans,  2017-04-13, LED is: 0, XI: 0x21
01605835 00000605 (2018-12-11 05:08:01.548641)  FW: Franz, 2018-12-05
01606231 00000396 (2018-12-11 05:08:01.944525)  FW: Hans,  2017-04-13, LED is: 0, XI: 0x21
01606836 00000605 (2018-12-11 05:08:02.549132)  FW: Franz, 2018-12-05
01607232 00000396 (2018-12-11 05:08:02.944913)  FW: Hans,  2017-04-13, LED is: 0, XI: 0x21
01607835 00000603 (2018-12-11 05:08:03.548483)  FW: Franz, 2018-12-05
01608232 00000397 (2018-12-11 05:08:03.945490)  FW: Hans,  2017-04-13, LED is: 0, XI: 0x21
01608836 00000604 (2018-12-11 05:08:04.549163)  FW: Franz, 2018-12-05
01609233 00000397 (2018-12-11 05:08:04.946070)  FW: Hans,  2017-04-13, LED is: 0, XI: 0x21
01609709 00000476 (2018-12-11 05:08:05.422676)  floppyEncoder_handleSaveFiles -- saving slot 0
```

**Comment by atarijookie** [comment]
Floppy write works - mostly, the write-read-verify test fails in 2 cases out of 50 (or 500, depends on point of view), looks like this (crc error):
```
02181896 00000012 (2018-12-14 04:10:06.349372)  handleSectorWritten -- track 22, side 0, sector 8
02181896 00000000 (2018-12-14 04:10:06.349512)  floppyEncoder_decodeMfmWrittenSector - track: 22, side: 0, sector: 8, size: 1062 -> writtenSector index: 0
02181897 00000001 (2018-12-14 04:10:06.350414)  MfmCachedImage::handleDecodedByte - received CRC: df37, calculated CRC: df36, good: 0
02181897 00000000 (2018-12-14 04:10:06.350543)  Written sector at index: 0 decoded, good: 0
```

On 427 kB file it had 10 empty / corrupted sectors (5 kB) -- 1.2% error... Good enough for science [EEEEEEEEE] Not Aperature Science! ;)

TODO:
- [ ] **investigage / fix CRC error from above**
- [x] add log ERROR on good: 0, so in normal (not debug) mode we will see only errornous writes
- [x] find out what byte combination produces the larges MFM stream per sector - to find out if 1100 bytes of write buffer is enough -- all ) or 1 should produce the longest stream (1B gets encoded into 2 B), so with headers and GAP this should be 1071 B (should fit in 1100 B write buffer)
- [x] use the same sector for whole track generation - to find out how much is really needed for whole track (now it's 15 kB -- could be bit lower, then the write buffer could be bit bigger) - calculated 12484 B maximum, some image showed 13205 B though
- [x] remove annoying overlay on encoding, which shouldn't now stop user from further action
- [ ] floppy save in CE_FDD.PRG not enabled
- [ ] floppy format ends up showing lots of failed-to-find-sync, but I'm not sure if that's an issue - we might not support this one, it's not needed when you can create working image using tools
- [ ] handle format track by checking the received streaming index and comparing it to valid 1st byte index, then erasing the whole track
- [ ] linear read test shows lazy reads on side 1 (side 0 is without any issue) -- when reading side 0, the SIDE1 signal is constantly H; but when reading size 0 the SIDE1 signal pulses 0-1-0 for 25 us from time to time, possibly making Franz want to stream the other side for a while...

**Comment by atarijookie** [comment]
- Read track resized from 15000 bytes to 13800 bytes. 
- Write sector resized from 550 to 650 words (1100 B -> 1300 B).

```
// Most of the normal games have tracks from 11000 bytes to 12019 bytes of encoded stream.
// With Bad_Brew_Crew/BBC11.ZIP (11 sectors per track?) the resulting stream was 13205 bytes big, I didn't see anything more yet (but there might be something bigger).
// In theory the 11 sectors per track could be up to 13720 bytes big, so 13800 should be enough here.
// In case something bigger appears (e.g. 12 sectors per track), consider shortening GAP 2, GAP 3a, GAP 3B, GAP 4, which take 774 bytes before encoding, 1548 bytes after encoding,
// so if more space for real data would be needed, some adaptive reducing of GAPs sizes could be added (to a point where WD1772 still considers those GAPs to be valid to catch the sync there)
#define MFM_STREAM_SIZE 13800
```

- old mfm encoder removed
- mfm track buffer protected against overflow
- encoding overlay removed from web, it shouldn't block users anymore

**Comment by atarijookie** [comment]
Possible reasons why 1.2% of write sectors are failing:
- Franz misses a pulse due to being under a load for a while
- some captured pulse is too long or too short and gets ignored
- if that would be the case, then we could disable interrupts during capturing, or go back to HW capturing
- cd ce_franz_fw ; git diff 069da82e64...ad043b05ef . (to see the HW capturing code from past)
- create a load on Franz by attaching injector and moving mouse and typing during the write
- measure how long the int handling takes to get the idea
- write buffer could be too short in some case (shouldn't happen anymore as the write buffer was resized from 550 w to 650 w)
- some last bit(s) is not stored to buffer before sending it to RPi
- that last bit is preceded by a space (empty byte) in the buffer, which RPi considers as the end of stream and doesn't add it to the decoded stream
- for testing - set only counter data to write buffer (no random part), set STOP after error, single sector write, and on RPi do data dump of decoded buffer after failure, look for corrupted data
- something else?

**Comment by atarijookie** [comment]
Can't make hw capturing work - captures around 680 B while sw capturing captures around 810 B...

- Find out, where the corrupted data in sw capturing is - start, middle, end - tests showed that the data is OK, there's only difference in CRC

- mark (counter?) how many pulses on write were too short or too long (less than 4 us, more than 8 us) - might help to understand why the data was wrong...

- Capturing loop profiling should be done...

- scope on wdata and wgate, trigger on wgate rise, see how it looks normally vs on error - no difference, plus there seem to be 6 pulses of 4 us after CRC, so we're probably not missing a capture of last CRC bit

- disable hw capture dma using #ifdef

**Comment by atarijookie** [comment]
With stream length determination + next sector index determination from stream table (instead of using a loop to find it manually) the situation didn't improve, and the CRC on bad write is still off just by one bit:
```
MfmCachedImage::handleDecodedByte - received CRC: 2acf, calculated CRC: 2ace, good: 0
MfmCachedImage::handleDecodedByte - received CRC: fadf, calculated CRC: fade, good: 0
```

**Comment by atarijookie** [comment]
- the data corruption seems to be only on CRC part (last 2 words) of the captured stream
- when this happens (1.2% of written sectors) the difference in calculated and received CRC is only 1 (0x2acf vs 0x2ace, 0xfadf vs 0xfade)
- when this error is considered as OK (accepting this difference), the write-read-compare test went OK through the whole floppy without single error (so all the read data was the same as written data)
- the scope shows, that when WGATE goes high, there seem to be at least 6 pulses of 4us after the CRC, so the last bit of CRC shouldn't be left uncaptured (this shouldn't be the cause of bad CRC received)

So my guess then would be that this is caused by:
- bad capturing of CRC in Franz
- bad decoding of MFM stream of CRC in RPi
- bad CRC calculation (missing one last bit?)

Conclusions:
- write-read-verify test using TOS functions, writing and reading 9 sectors at once, passed with 1440 W-R-V ops when accepting the CRC off by one (so data is still valid)
- W-R-V test using custom functions with 9 sectors at once fails always, there seems to be a bug there...
- W-R-V test using custom functions with 1 sector at once had 5 errors on 1440 sectors (0.35%), which had a correct CRC on RPi when received, so this will be most likely caused by not streaming back the written sector and ST reading still the old content (doesn't happen with TOS functions, as they probably are slower)
- copying 427 kB file from shared drive to floppy and back showed no difference in content (compare by content on pc)

**Comment by atarijookie** [comment]
TODO:
- [x] fix W-R-V with custom functions and multiple sectors
- [ ] check what causes fail on W-R-V with custom functions on single sector (lazy read?) -- it seems that the new data hasn't been sent yet to Franz at the moment the ST tries to read it back
- [x] fix lazy read on linear floppy read test
- [x] enable floppy saving in floppy PRG
- [x] test floppy saving through ce_fdd.prg
- [ ] add detection of offset to sector stream table - it will be most probably 5 empty words most of the time, but it might be 6 when something extra is received, so don't rely on 10 byte offset, but do a check for offset instead
- [x] verify and disable writing to config image and test image (or just re-generate test image on insert, so it can be written, but if doing read only test, if will be still valid)
- [x] implement floppy format - handle ID MARK found when DAM mark not found
- [x] check mouse / keyboard reliability through injector while doing floppy write - mouse movements stutter during write, but the written data is OK
- [ ] refactor IKBD receiving / sending - (see refactoring options bellow) read uart and store to buffer, send to ST when not writing
- [ ] mouse movement in TOS when deleting file (writing to floppy) causes keyboard / mouse misbehave as some IKBD data is disrupted :(
- [ ] mouse movement in floppy test while doing write leads to bombs after a while (50 - 100 writes) as probably the corrupted IKDB data isn't handled well in TOS and ends up like that
- [ ] returning from CE_FDD.PRG to desktop causes TOS windows unresponsive - memory corruption, or failed to release resources?

**Comment by atarijookie** [comment]
While working with all of this, I've also hit another issue:
- when ce_main_app runs with noreset argument, Franz doesn't want to start with HSE, ends up running on 8 MHz and won't communicate with RPi 👎 
- when ce_main_app runs normally (and does Franz's reset), the HSE starts, ends up running at 72 MHz as it should be 👍 

...so I've added a code, that if the external clock (HSE) doesn't want to start, use internal clock (HSI), with which it ends up running at 64 MHz. I have no explanation yet for this behavior, but the timer prescalers and write capturing has been altered to take in mind this possibility of running at either 72 MHz or 64 MHz. 🥇 

Tested floppy reading and writing and also IKBD injection while running at 72 MHz and 64 MHz and both were running fine. 👍

**Comment by atarijookie** [comment]
**IKBD refactoring options:**
**A)** simple but might not work - either let IRQ handlers receive UARTs during write 

**B)** simple but might not work -- c&p UART receiving to floppy_write() function. Possibly handle only UART3 RX (to slow down floppy write as least as possible), plus use some local buffer for faster storage of received bytes (circular buffer might be too slow here), and put all the received bytes from local buffer to the right cyclic buffer. 
The checking and handling of received UART byte should be possibly done _AFTER_ detecting the low WDATA pulse, as you can easily miss the WDATA pulse as it might be only 0.5 us wide, but after detecting that the pulse was there you have 4/6/8 us until the next pulse - with the WDATA processing the 6 and 8 us pulse would be enough for UART processing, let's think of something for the 4 us wide pulse...
For changes to make the WDATA going low detection not being such time critical, we might either use HW pulse capturing, or falling event capturing (EXTI)... 

**C)** more complicated, but can't be done due to already used up DMA channels - alter UART receiving to use 3x DMA into circular buffer, remove IRQ handlers for UART, add code for reading that DMA filled circular buffer and let it TX to the right UARTs outside of floppy_write() function()

Option **C)** can't be done -- the DMA1 of Franz has 7 channels, 3 are used already. It should be probably fine with just UART3 being received (this one comes from original IKBD), but UART3 RX is handled by DMA1 channel 3, which is already used for SPI1 TX (background transfer from Franz to RPi), so it even can't be reconfigured for temporary use for UART3...

**Comment by atarijookie** [comment]
Notes after today's tests:
- Timer Capture while handling one capture at a time doesn't work well - same as when using it with DMA - it captures ~720 pulses instead of ~820 pulses when doing it by GPIO checking - fixing this would also fix the DMA capture, which could possibly work with IKBD at the same time... 
- adding UART / FIFO handling after capture doesn't work with FIFO functions, which are slow, but seemed to work better with storing to linear buffer (only for UART3), but it ended up in crashed TOS anyway - maybe capturing also what comes from RPi might be needed
- tried to rewrite UART IRQ handlers to work faster, but they ended up not working at all, will have to check what's going on

To see why the Timer Capture returns different values, I could do the following:
- on pulse time captured - store the value into buffer, e.g. first 64 captured values
- write single sector, software capture, check values in buffer
- write single sector, timer capture, check values in buffer

By reading the sources of flashfloppy / gotek I see that they are running the Timer Capturing at full 72 MHz (0.0138 us period), while I have it pre-scaled to 8 MHz (0.125 us period). The write pulse low time is around 0.5 us, that would be 4 ticks @ 8 MHz or 36 ticks @ 72 MHz. Maybe sometimes the write pulse gets even narrower than that and edge detector misses it when running at this lower timer frequency. 

**TODO:** 
- [x] test that sw / timer capture values to buffer as mentioned above - pulses were too short, needed different values for classification into 4-6-8 us
- false -  is just input trigger selection missing? TIM_SelectInputTrigger(TIM3, TIM_TS_TI1F_ED);
- not needed - test capturing at full 72 MHz timer clock instead of 8 MHz
- not happening - add checking + counter of overcapture happening: TIM3->SR -- CC1OF: overcapture flag

**Comment by atarijookie** [comment]
Finally some positive results - I've managed to get the Timer Input Capture to work - it was just bad processing of data and ignoring the short pulses... I still need to figure out how to solve the IKBD receiving and sending while doing floppy write...

I can't send the UART data only in main read loop - when doing write of 9 sectors in a row, the buffers get full as they aren't emptied often enough. This was tested by capturing the UART1 and UART3 to linear buffer during sector write and emptied to circular buffer after sector write. 

What can be done:
- [ ] improve UART IRQ handlers to take less time, so they possibly don't screw up either written data or IKDB data
- [ ] fix / finish DMA capturing now that the timer input capture works. This could leave enough time between DMA handlings to process the UARTs (IKBD) normally in IRQ handlers

**Comment by atarijookie** [comment]
After few months I've made few improvements and tests, and the results are as before:
- SW capturing results in good floppy data, but bad mouse behavior during write
- HW capturing results in good mouse behavior during write, but bad floppy data

I was thinking of HW (DMA) UART handling, but that's not possible with the current UARTs vs free DMAs.

There's also another option for partial HW capture:
- instead of Timer Capture + DMA for storage, we still could try Timer Capture + interrupt on timer capture. But if the issue is in Timer Capture, I'm not sure if this interrupt on capture would solve anything.

The worst possible, but still working solution would be:
- use SW capture which results in good data, but tell people to not move mouse around during floppy write (which would be annoying / not comfortable).

I've also stored the count of occurrence of captured duration through timer and DMA on a single sector, and it's like this:
- 0 ticks duration - 32 times - okay, this is fine, another capture after the short one will capture the whole length
- 28-32 ticks - 3.5 - 4.0 us - fine
- 47-51 ticks - 5.8 - 6.4 us - fine
- 63-67 ticks - 7.8 - 8.4 us - fine
- **ticks too long - 533 times - this is too bad...** happens when the timer capture misses a low pulse, this can't be fixed by sw, need to investigate this, or go with some other solution

**Comment by atarijookie** [comment]
TODO:
- TIM3 CH1 period seems to be wrong - TIM_TimeBaseStructure.TIM_Period = 0xff; - we let it go up to 0xff, but store it and work with it as with WORD - this seems very wrong ;)
- store count of occurrence of captured durations, but don't limit to 100, but rather to 200 or 300, to see the real length of 'ticks too long'
- try leaving DMA on, don't do DMA off, DMA on and so on
- set higher priority on DMA write channel
- disable DMA read channel while doing DMA write

**Comment by atarijookie** [comment]
Okay, so it was TIM_Period = 0xff; instead of 0xffff; -- floppy write now works with HW capturing and you can move the mouse during that and nothing bad happens :-D

I'm closing this task as the floppy write has been now added and merged to master, any other related tasks or bugs might get an own issue. 

I don't want to continue in this issue, it's too long and depressing :-D (create new issues for bugs / todo related to floppy / floppy write).
